pyquil/api/qpu.py

```diff
@@ -13,27 +13,99 @@
 
 from pyquil.quil import Program
 from ._base_connection import validate_run_items, TYPE_MULTISHOT, TYPE_MULTISHOT_MEASURE, get_job_id, BaseConnection
+from ._config import PyquilConfig
+
+
+def get_devices(async_endpoint='https://job.rigetti.com/beta', api_key=None, user_id=None):
+    """
+    Get a list of currently available devices. The arguments for this method are the same as those for QPUConnection.
+    Note that this method will only work for accounts that have QPU access.
+
+    :return: list of online and offline devices
+    :rtype: list
+    """
+    config = PyquilConfig()
+    api_key = api_key if api_key else config.api_key
+    user_id = user_id if user_id else config.user_id
+
+    headers = {
+        'X-Api-Key': api_key,
+        'X-User-Id': user_id,
+        'Content-Type': 'application/json; charset=utf-8'
+    }
+
+    response = requests.get(async_endpoint + '/devices', headers=headers)
+    return {Device(name, device) for (name, device) in response.json()['devices'].items()}
+
+
+class Device(object):
+    """
+    A device (quantum chip) that can accept programs. Only devices that are online will actively be accepting new
+    programs.
+    """
+    def __init__(self, name, raw):
+        """
+        :param name: name of the device
+        :param raw: raw JSON response from the server with additional information about this device
+        """
+        self.name = name
+        self.raw = raw
+
+    def is_online(self):
+        """
+        Whether or not the device is online and accepting new programs.
+        :rtype: bool
+        """
+        return self.raw['is_online']
+
+    def __str__(self):
+        online_offline = 'online' if self.is_online() else 'offline'
+        return '<Device {} {}>'.format(self.name, online_offline)
+
+    def __repr__(self):
+        return str(self)
 
 
 class QPUConnection(BaseConnection):
     """
     Represents a connection to the QPU (Quantum Processing Unit)
     """
 
-    def __init__(self, async_endpoint='https://job.rigetti.com/beta', api_key=None, user_id=None):
+    def __init__(self, device_name=None, async_endpoint='https://job.rigetti.com/beta', api_key=None, user_id=None):
         """
         Constructor for QPUConnection. Sets up necessary security.
 
         :param async_endpoint: The endpoint of the server for running QPU jobs
         :param api_key: The key to the Forest API Gateway (default behavior is to read from config file)
         :param user_id: Your userid for Forest (default behavior is to read from config file)
         """
+        if not device_name:
+            warnings.warn("""
+You created a QPUConnection without specificying a device name. This means that
+your program will be sent to a random, online device. This is probably not what
+you want. Instead, pass a device name to the constructor of QPUConnection:
+
+    qpu = QPUConnection('the_name')
+
+To get a list of available devices, use the get_devices method, for instance:
```

Comment:
should we add a note about permissions? not all users will get anything useful from `get_devices` but any user who instantiates this class could see this message